tests/lib/fakeinstaller/main.go

```diff
@@ -0,0 +1,282 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/gadget"
+	"github.com/snapcore/snapd/gadget/install"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/osutil/disks"
+	"github.com/snapcore/snapd/osutil/mkfs"
+)
+
+func firstVol(volumes map[string]*gadget.Volume) *gadget.Volume {
+	for _, vol := range volumes {
+		return vol
+	}
+	return nil
+}
+
+func createPartitions(bootDevice string, volumes map[string]*gadget.Volume) ([]gadget.OnDiskStructure, error) {
+	// TODO: support multiple volumes, see gadget/install/install.go
+	if len(volumes) != 1 {
+		return nil, fmt.Errorf("got unexpected number of volumes %v", len(volumes))
+	}
+
+	diskLayout, err := gadget.OnDiskVolumeFromDevice(bootDevice)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read %v partitions: %v", bootDevice, err)
+	}
+	if len(diskLayout.Structure) > 0 {
+		return nil, fmt.Errorf("cannot yet install on a disk that has partitions")
+	}
+
+	layoutOpts := &gadget.LayoutOptions{
+		IgnoreContent: true,
+	}
+
+	vol := firstVol(volumes)
+	lvol, err := gadget.LayoutVolume(vol, gadget.DefaultConstraints, layoutOpts)
+	if err != nil {
+		return nil, fmt.Errorf("cannot layout volume: %v", err)
+	}
+
+	iconst := &install.CreateOptions{CreateAllMissingPartitions: true}
+	created, err := install.CreateMissingPartitions(diskLayout, lvol, iconst)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create parititons: %v", err)
+	}
+	logger.Noticef("created %v partitions", created)
+
+	return created, nil
+}
+
+func runMntFor(label string) string {
+	return filepath.Join(dirs.GlobalRootDir, "/run/fakeinstaller-mnt/", label)
+}
+
+func postSystemsInstallSetupStorageEncryption(details *client.SystemDetails) error {
+	// TODO: check details.StorageEncryption and call POST
+	// /systems/<seed-label> with "action":"install" and
+	// "step":"setup-storage-encryption"
+	return nil
+}
+
+// XXX: reuse/extract cmd/snap/wait.go:waitMixin()
+func waitChange(chgId string) error {
+	cli := client.New(nil)
+	for {
+		chg, err := cli.Change(chgId)
+		if err != nil {
+			return err
+		}
+
+		if chg.Err != "" {
+			return errors.New(chg.Err)
+		}
+		if chg.Ready {
+			return nil
+		}
+		time.Sleep(1 * time.Second)
+	}
+}
+
+// TODO laidoutStructs is used to get the devices, when encryption is
+// happening maybe we need to find the information differently.
+func postSystemsInstallFinish(cli *client.Client,
```

Comment:
(nitpick) linewrapping at column 49 ? ;) (I will shut up about the long lines vs short lines, hope I'm not already getting on your nerves)

Reply:
Looking forward to our next sprint and your setup, I will probably change my mind then :)